File: 10. Backtracking/Combinations/sol1.cpp
#include <iostream>
#include <string>
#include <stdio.h>
#include <ctype.h>
#include <vector>
using namespace std;

void combineHelper(vector<int> rest,vector<int> soFar,int k, vector<vector<int>> &result){
  if (k==0){
    result.push_back(soFar);
  }
  else if ((rest.empty()) || (k>rest.size())){
    return;
  }
  else{
    vector<int> soFar_next=soFar;
    soFar_next.push_back(rest[0]);
    vector<int> rest_next=rest;
    rest_next.erase(rest_next.begin()+0);
    combineHelper(rest_next,soFar_next,k-1,result);
    combineHelper(rest_next,soFar,k,result);
  }
}

vector<vector<int> > combine(int n, int k) {
  vector<int> rest;
  vector<int> soFar;
  for (int i=1;i<=n;i++){
    rest.push_back(i);
  }
  vector<vector<int> > result;
  combineHelper(rest,soFar,k,result);
  return result;
}

int main(int argc, const char * argv[]) {
  int n=4;
  int k=2;
  vector <vector <int>> out=combine(n,k);
  for (int i=0;i<out.size();i++){
    for (int j=0;j<out[i].size();j++){
      cout<<out[i][j]<<" ";
    }
    cout<<endl;
  }
  cout<<endl;
  return 0;
}

File: 10. Backtracking/Combinations/sol1.py
class Solution:
    # @param A : list of integers
    # @return a list of list of integers
    def __init__(self):
        self.result=[]
    def subsetsHelper(self,soFar,Rest,k ):
        print (soFar,Rest,k)
        if  (k==0):
            print ("yo ",soFar,k)
            self.result.append(sorted(soFar))
            return 
        elif ((Rest is None) or (k>len(Rest))):
            print ("here")
            return
        else:
            self.subsetsHelper(soFar+[Rest[0]],Rest[1:],k-1)
            self.subsetsHelper(soFar,Rest[1:],k)

    def combine(self, A, k):
        Rest=[i for i in range(1,A+1)]
        self.subsetsHelper([],Rest,k)
        return sorted(self.result)
sol_obj=Solution()
A=4
k=2
out=sol_obj.combine(A,k)
print (out)
